### PR TITLE
XCM revamp & Ping pallet

### PR DESCRIPTION
**Pull request by gavofyork**
Depends on https://github.com/paritytech/polkadot/pull/2836
Tracked by paritytech/polkadot#2841

Also includes XCMP message dispatch system reimagining:
- XCM messages now get aggregated into an XCMP message before being sent out
- Similarly, XCMP messages get split into XCM messages before being executed
- Moved most of the logic into xcm-handler pallet
- Altered the outgoing XCMP API from push to pull
- Changed underlying outgoing queue data structures to avoid multi-page read/writes
- Introduced queuing for incoming messages
- Introduced signal messages as a flow-control sub-stream
- Introduced flow-control with basic threshold back-pressure
- Introduced overall weight limitation on messages executed
- Additonal alterations to XCM APIs for the new system

Also introduces the ping pallet for pinging between chains.

### CODE REVIEW

File: pallets/xcm-handler/src/lib.rs [hunk 1]
```diff
@@ -74,134 +128,662 @@ decl_error! {
...
+						.map_err(|e| XcmError::CannotReachDestination(<&'static str>::from(e)))?;
+					Self::deposit_event(RawEvent::XcmpMessageSent(Some(hash)));
 					Ok(())
 				} else {
 					Err(XcmError::UnhandledXcmMessage)
 				}
 			}
 			_ => {
-				/* TODO: Handle other cases, like downward message */
+				// Here is where we handle additional cases such as into parachains, bridges or
```
Comment (gavofyork) [inline pallets/xcm-handler/src/lib.rs:785]: CC @xlc

File: pallets/xcmp-queue/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+	/// dispatched in order, then both must be sent with `ServiceQuality::Ordered`.
+	///
+	/// ## Background
+	///
+	/// For our purposes, one HRMP "message" is actually an aggregated block of XCM "messages".
+	///
+	/// For the sake of clarity, we distinguish between them as message AGGREGATEs versus
+	/// message FRAGMENTs.
+	///
+	/// So each AGGREGATE is comprised af one or more concatenated SCALE-encoded `Vec<u8>`
```
Comment (4meta5) [inline pallets/xcmp-queue/src/lib.rs:200]:
```suggestion
	/// So each AGGREGATE is comprised of one or more concatenated SCALE-encoded `Vec<u8>`
```

File: pallets/xcm/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+	}
+
+	#[pallet::event]
+//	#[pallet::generate_deposit(pub(super) fn deposit_event)]
```
Comment (bkchr) [inline pallets/xcm/src/lib.rs:45]:
```suggestion
```

File: pallets/xcm/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
```
Comment (bkchr) [inline pallets/xcm/src/lib.rs:38]:
```suggestion
	/// The module configuration trait.
	#[pallet::config]
```

File: rococo-parachains/pallets/ping/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
```
Comment (bkchr) [inline rococo-parachains/pallets/ping/src/lib.rs:40]:
```suggestion
	/// The module configuration trait.
	#[pallet::config]
```

File: rococo-parachains/pallets/ping/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		type Origin: From<<Self as SystemConfig>::Origin> + Into<Result<CumulusOrigin, <Self as Config>::Origin>>;
+
+		/// The overarching call type; we assume sibling chains use the same type.
+		type Call: From<Call<Self>> + Encode;
+
+		type XcmSender: SendXcm;
+	}
+
+	#[pallet::storage]
+	/// Details of an asset.
```
Comment (bkchr) [inline rococo-parachains/pallets/ping/src/lib.rs:54]:
```suggestion
	/// The target parachains to ping.
	#[pallet::storage]
```

File: rococo-parachains/pallets/ping/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	pub(super) type Targets<T: Config> = StorageValue<
+		_,
+		Vec<(ParaId, Vec<u8>)>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	/// Details of an asset.
```
Comment (bkchr) [inline rococo-parachains/pallets/ping/src/lib.rs:62]:
```suggestion
	/// The total number of pings sent.
	#[pallet::storage]
```

File: rococo-parachains/pallets/ping/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+	pub(super) type PingCount<T: Config> = StorageValue<
+		_,
+		u32,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	/// Details of an asset.
```
Comment (bkchr) [inline rococo-parachains/pallets/ping/src/lib.rs:70]:
```suggestion
	/// The sent pings.
	#[pallet::storage]
```

File: primitives/utility/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helper datatypes for cumulus. This includes the `ParentAsUmp` routing type which will route
+//! messages into an `UpwardMessageSender` if the destination is `Parent`.
```
Comment (bkchr) [inline primitives/utility/src/lib.rs:18]:
```suggestion
//! Helper datatypes for cumulus. This includes the [`ParentAsUmp`] routing type which will route
//! messages into an [`UpwardMessageSender`] if the destination is `Parent`.
```

File: primitives/utility/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+			.map(Xcm::<Call>::try_from);
+		let weight_used = match msg {
+			Ok(Ok(x)) => {
+				let weight_limit = MaxWeight::get();
+				XcmExecutor::execute_xcm(Junction::Parent.into(), x, weight_limit).weight_used()
+			}
+			Ok(Err(..)) => 0,
+			Err(..) => 0,
+		};
+		weight_used
```
Comment (bkchr) [inline primitives/utility/src/lib.rs:76]:
```suggestion
		match msg {
			Ok(Ok(x)) => {
				let weight_limit = MaxWeight::get();
				XcmExecutor::execute_xcm(Junction::Parent.into(), x, weight_limit).weight_used()
			}
			Ok(Err(..)) => 0,
			Err(..) => 0,
		}
```

File: pallets/parachain-system/src/lib.rs [hunk 11]
```diff
@@ -292,109 +329,19 @@ decl_module! {
...
+				maximum_channels,
+			);
 
-			// Sort the outbound messages by asceding recipient para id to satisfy the acceptance
-			// criteria requirement.
-			outbound_hrmp_messages.sort_by_key(|m| m.recipient);
+			// Note conversion to the OutboundHrmpMessage isn't needed since the data that
+			// `take_outbound_messages` returns encodes equivalently.
+			// If the following code breaks, then we'll need to revisit that assumption.
+			let _ = OutboundHrmpMessage { recipient: ParaId::from(0), data: vec![] };
```
Comment (shawntabrizi) [inline pallets/parachain-system/src/lib.rs:342]: whats happening here?

File: pallets/xcm/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+	}
+
+	#[pallet::event]
+//	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T> = "Balance", AssetIdOf<T> = "AssetId")]
```
Comment (shawntabrizi) [inline pallets/xcm/src/lib.rs:46]:
```suggestion
```

File: pallets/xcmp-queue/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,706 @@
...
+			InboundXcmpMessages::remove(sender, sent_at);
+		} else {
+			InboundXcmpMessages::insert(sender, sent_at, remaining_fragments);
+		}
+		(weight_used, is_empty)
+	}
+
+	/// Service the incoming XCMP message queue attempting to execute up to `max_weight` execution
+	/// weight of messages.
+	fn service_xcmp_queue(max_weight: Weight) -> Weight {
```
Comment (shawntabrizi) [inline pallets/xcmp-queue/src/lib.rs:409]: Is the behavior in this function spec'ed somewhere?

File: rococo-parachains/pallets/ping/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		fn pong(origin: OriginFor<T>, seq: u32, payload: Vec<u8>) -> DispatchResult {
+			// Only accept pings from other chains.
+			let para = ensure_sibling_para(<T as Config>::Origin::from(origin))?;
+
+			if let Some(sent_at) = Pings::<T>::take(seq) {
+				Self::deposit_event(Event::Ponged(para, seq, payload, frame_system::Pallet::<T>::block_number() - sent_at));
```
Comment (shawntabrizi) [inline rococo-parachains/pallets/ping/src/lib.rs:184]:
```suggestion
				Self::deposit_event(Event::Ponged(para, seq, payload, frame_system::Pallet::<T>::block_number().saturating_sub(sent_at)));
```

File: rococo-parachains/pallets/ping/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+		_,
+		Blake2_128Concat,
+		u32,
+		T::BlockNumber,
+		OptionQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T> = "Balance", AssetIdOf<T> = "AssetId")]
```
Comment (shawntabrizi) [inline rococo-parachains/pallets/ping/src/lib.rs:81]:
```suggestion
	#[pallet::metadata(T::BlockNumber = "BlockNumber")]
```

**Comment by shawntabrizi** [comment]
`pallets/xcmp-queue/src/lib.rs` needs a ton of tests and some specification on how it should behave.

Some really specific logic in it.

File: pallets/parachain-system/src/lib.rs (hunk 11 above)
Comment (gavofyork) [reply pallets/parachain-system/src/lib.rs:342]: It's a smoke test checking that the fields of `OutboundHrmpMessage` have not changed.

**Review by shawntabrizi — APPROVED** [review]
I think it should be fine to merge if there is a new issue opened to add tests particularly to the `xcmp-queue`

**Pull request merged**